Add events needed for replication

cowlicks opened the pull request:
This adds some events to `Hypercore` that are needed for replication to work.

It also defines some traits that can get used in place of a concrete `Hypercore`. So downstream libraries can just consume a `impl CoreMethods` which will work for both `Hypercore` and `Arc<Mutex<Hypercore>>`.

I also fixed some TODO's and small lints I came across in the docs.

One thing that may be controversial:
The `replication` feature requires `tokio` because it defines a `SharedCore` type (that uses `tokio::sync::Mutex`) which is intended to be the de-facto implementation for downstream libraries to use when they want a `Hypercore` that can have multiple owners. I use this downstream in the [`replicator`](https://github.com/cowlicks/replicator) crate.

In theory, the dependence on `tokio` could be removed, by moving `SharedCore` to it's own crate. And the other usage of `tokio::sync::broacast` could be replaced with [`async_broadcast`](https://github.com/smol-rs/async-broadcast/tree/master) which is async runtime agnostic. However this seemed like it could be more overhead than it is worth.

cowlicks commented on src/core.rs:
```diff
@@ -330,10 +342,32 @@ impl Hypercore {
+            // if not in this core, try to get over network
+            {
+                let mut rx = self.events.send_on_get(index);
+                tokio::spawn(async move {
```
I think that would make sense, but should probably be configurable like the [js get method is](https://github.com/datrs/hyperswarm-rs/pull/3), to have `wait` and `timeout` parameter. However changing this methods signature would be a breaking change. So I didn't want to do it in this PR.

The default behavior in js seems to be `wait = true` but `timeout = 0`. Which effectively does the same thing here, it notifies the replicator that we want to get a block, but does not wait on it. Relevant JS [here](https://github.com/holepunchto/hypercore/blob/7dcd252d281dfc3f2dd62260d27573901791c4b6/index.js#L897-L908)

I propose adding a separate `Hypercore::get_opt(index: u64, opts: GetOpts) -> Future<Result<...>>` but doing so in a separate PR.

RangerMauve replied:
Ah I see. So with this code the behavior would look like:

```rust
if core.get(1337).is_none() {  // triggers download in background
// some time later, do we wait for an event?
core.get(1337) -> Some(bytes) // got downloaded!
}
```

It feels like it'd be better to decouple `try get + trigger download` into separate things. Like maybe `get()` would return `None` always until you trigger a download via `download(index|range).await`.

Then code would look like

```rust
if core.get(1337).is_none() {
  core.download(1337).await?;
  return core.get(1337);
}

cowlicks replied:
The current behavior matches JS, I think diverging from that could be confusing to users. I'm unsure about adding a `Hypercore::download` method at this time since it would **require** the `replication` feature and would be harder to do in a async runtime agnostic way.

cowlicks replied:
I was thinking about this before. There is actually no need for the `spawn`. This should just send the event that we tried to get. It is up to the subscriber for how to handle it. But in the case of the replicator implementation, the behavior is the same as JS.

ttiurani commented:
Thanks for the PR, overall looks reasonable!

However, there are no tests for it here now at all, and that makes it fragile to maintain. Would you consider adding tests?

Wrt the async_std support, this is quite a big decision. After so much effort in having multi-async support, just casually dropping it is kinda a massive decision? Personally I'd **either** drop async_std everywhere consistently (random-access-disk and hypercore-protocol-rs) and make it all tokio-only, **or** then put in the effort to make it work in both.

Even though `async-std` might get deprecated in the coming years, I can see hypercore being useful in some `smol` environments, and swapping those async runtimes is doable if we don't do a tokio-only features here.

Could you try to use `async_broadcast` to see what that would look like?

cowlicks commented:
Good point about the tests! I forgot about that here because I tested it in the `replicator` crate. However it'd be good to have tests here too.

I'll work on adding tests and de-coupling `tokio` & `replication` here.

cowlicks commented:
I managed to completely decouple tokio from the replication stuff using `smol`'s crates which are runtime agnostic (It may be possible to do the same thing within the random-access-* crates which would let us drop the `tokio` & `async-std` features completely so that we could be completely runtime agnostic). It turned out I did not need `tokio` for `SharedCore` so there was no need to remove it.

I also added a few tests. There are probably more needed but I am done for the day.

cowlicks commented:
I moved the `SharedCore` behind it's own feature (`shared-core = ["replication", "dep:async-lock"]`).

I see we test, almost every combination of features, and instead of multiplying the number of tests runs by x12 (for the two new features x (windows, mac, linux)) I added just one `cargo test --no-default-features --features tokio,shared-core` to each target. These are added before the existing & failing `js_interop_tests` so you can see they pass in the CI.

An aside: the `async-std` and `tokio` features only effect dependencies. So testing with both of these seems like we are testing implementation details of our own dependencies. It may be better to leave that to those dependencies to test themselves.

RangerMauve reviewed (approved):
LGTM

ttiurani commented on .github/workflows/ci.yml:
```diff
@@ -38,6 +38,7 @@ jobs:
         cargo check --no-default-features --features async-std
         cargo check --no-default-features --features async-std,sparse
         cargo check --no-default-features --features async-std,sparse,cache
+        cargo test --no-default-features --features tokio,shared-core
```
Any reason you aren't adding also async-std replication tests here too, instead of these tokio-only tests?

cowlicks replied:
I've added the `async-std` tests.

ttiurani reviewed (approved):
Fantastic work! Looks perfect now.

Thank you for fixing also TODOs along the way.
